File: cpp_drill/Problems/TR13_CountGoodNodesInBT.cpp
#include <iostream>
using namespace std;

// REF : https://leetcode.com/problems/count-good-nodes-in-binary-tree/
// NOTES :
//      Apply DFS and in each call if node val is > than it's ascendants
//      then update the max_val with node->val and increase the count.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// T: O(n)
// S: O(n)
class Solution {
public:
    int goodNodes(TreeNode* root) {
        int count = 0;
        dfs(root, root->val, count);
        return count;
    }

    void dfs(TreeNode* node, int max_val, int& count) {
        if (!node)
            return;

        if (node->val >= max_val) {
            count++;
            max_val = node->val;
        }

        if (node->left)
            dfs(node->left, max_val, count);

        if (node->right)
            dfs(node->right, max_val, count);
    }
};

File: py_drill/Problems/TR13_CountGoodNodesInBT.py
# REF : https://leetcode.com/problems/count-good-nodes-in-binary-tree/
# NOTES :
#      Apply DFS and in each call if node val is > than it's ascendants
#      then update the max_val with node.val and increase the count.

from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# T: O(n)
# S: O(n)
# DFS recursive
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        if root is None:
            return 0

        def dfs(node, maxVal):
            nonlocal count
            if node.val >= maxVal:
                count += 1

            if node.left:
                dfs(node.left, max(node.val, maxVal))
            if node.right:
                dfs(node.right, max(node.val, maxVal))

        count = 0
        dfs(root, float("-inf"))
        return count

# T: O(n)
# S: O(n)
# DFS iterative & BFS iterative with popleft change
class Solution2:
    def goodNodes(self, root: TreeNode) -> int:
        if root is None:
            return 0

        count = 0
        s = deque()
        s.append([root, float("-inf")])
        while s:
            node, maxVal = s.pop()          # BFS with popleft
            if node.val >= maxVal:
                count += 1
            if node.left:
                s.append([node.left, max(maxVal, node.val)])
            if node.right:
                s.append([node.right, max(maxVal, node.val)])

        return count


def main():
    pass

if __name__ == "__main__":
    main()